src/WebTelemetryBase.ts

```diff
@@ -95,36 +94,51 @@ export abstract class WebTelemetryBase<P, R> {
         }
     }
 
-    protected createEvent<M>(payload: P, meta?: M): WebTelemetryBaseEvent {
+    protected createEvent<M>(payload: P, meta?: M): Promise<WebTelemetryBaseEvent> {
         let evt: WebTelemetryBaseEvent = {
             sessionId: globalSessionId,
             ...this.payloadToJSON(payload),
         };
 
         let evtMetadata = meta || {};
 
-        for (const addon of this.addons) {
-            const data = addon.data();
-            const metadata = addon.metadata();
-            evt = Object.assign({}, data, evt);
-            evtMetadata = Object.assign({}, metadata, evtMetadata);
-        }
-
-        evt.metadata = stringifyCircularObj(evtMetadata);
-
-        return evt;
+        const addonPromises = this.addons.map((addon) => Promise.all([addon.data(), addon.metadata()]));
+
+        return Promise.all(addonPromises).then((results) => {
+            const [finalEvt, finalMetadata] = results.reduce(
+                ([currentEvt, currentMetadata], [dataResult, metadataResult]) => {
+                    return [
+                        {
+                            ...dataResult,
+                            ...currentEvt,
+                        },
+                        {
+                            ...metadataResult,
+                            ...currentMetadata,
+                        },
+                    ];
+                },
+                [{}, evtMetadata],
+            );
+
+            return {
+                ...{ ...evt, ...finalEvt },
+                metadata: stringifyCircularObj(finalMetadata),
+            };
+        });
     }
 
-    public push<M>(payload: P, meta?: M): WebTelemetryBaseEvent {
+    public push<M>(payload: P, meta?: M): Promise<WebTelemetryBaseEvent> | WebTelemetryBaseEvent {
         if (this.config.disabled) {
             return { sessionId: 'disabled' };
         }
```

Comment:
Лучше сделать так, чтобы метод всегда возвращал Promise. Так не будет соблазна его не обработать

```suggestion
    public push<M>(payload: P, meta?: M): Promise<WebTelemetryBaseEvent> {
        if (this.config.disabled) {
            return Promise.resolve({ sessionId: 'disabled' });
        }
```

Reply:
Да. Так будет лучше.